File: core/src/main/java/io/github/shaksternano/borgar/core/media/reader/JavaxImageReader.java
package io.github.shaksternano.borgar.core.media.reader;

import com.google.common.collect.Iterators;
import io.github.shaksternano.borgar.core.collect.ClosableIterator;
import io.github.shaksternano.borgar.core.collect.ClosableSpliterator;
import io.github.shaksternano.borgar.core.media.ImageFrame;
import io.github.shaksternano.borgar.core.media.ImageUtil;
import io.github.shaksternano.borgar.core.media.MediaReaderFactory;
import org.jetbrains.annotations.Nullable;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.function.Consumer;

public class JavaxImageReader extends BaseMediaReader<ImageFrame> {

    private final ImageFrame image;

    public JavaxImageReader(File input, String format) throws IOException {
        this(ImageIO.read(input), format);
    }

    public JavaxImageReader(InputStream input, String format) throws IOException {
        this(ImageIO.read(input), format);
        input.close();
    }

    private JavaxImageReader(@Nullable BufferedImage image, String format) throws IOException {
        super(format);
        if (image == null) {
            throw new IOException("Failed to read image");
        }
        // For some reason some images have a greyscale type, even though they have color
        var converted = ImageUtil.convertType(image, BufferedImage.TYPE_INT_ARGB);
        this.image = new ImageFrame(converted, 1, 0);
        frameCount = 1;
        duration = 1;
        frameRate = 1;
        frameDuration = 1;
        width = converted.getWidth();
        height = converted.getHeight();
    }

    @Override
    public ImageFrame readFrame(long timestamp) {
        return image;
    }

    @Override
    public ImageFrame first() {
        return image;
    }

    @Override
    public MediaReader<ImageFrame> reversed() {
        return this;
    }

    @Override
    public ClosableIterator<ImageFrame> iterator() {
        return ClosableIterator.wrap(Iterators.singletonIterator(image));
    }

    @Override
    public void forEach(Consumer<? super ImageFrame> action) {
        action.accept(image);
    }

    @Override
    public ClosableSpliterator<ImageFrame> spliterator() {
        return ClosableSpliterator.wrap(List.of(image).spliterator());
    }

    @Override
    public void close() {
    }

    public enum Factory implements MediaReaderFactory<ImageFrame> {

        INSTANCE;

        @Override
        public MediaReader<ImageFrame> createReader(File media, String format) throws IOException {
            return new JavaxImageReader(media, format);
        }

        @Override
        public MediaReader<ImageFrame> createReader(InputStream media, String format) throws IOException {
            return new JavaxImageReader(media, format);
        }
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/reader/JavaxImageReader.kt
package io.github.shaksternano.borgar.core.media.reader

import io.github.shaksternano.borgar.core.collect.CloseableIterator
import io.github.shaksternano.borgar.core.collect.CloseableSpliterator
import io.github.shaksternano.borgar.core.io.DataSource
import io.github.shaksternano.borgar.core.media.ImageFrame
import io.github.shaksternano.borgar.core.media.ImageReaderFactory
import io.github.shaksternano.borgar.core.media.convertType
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.awt.image.BufferedImage
import java.io.IOException
import java.util.function.Consumer
import javax.imageio.ImageIO
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds

class JavaxImageReader(
    image: BufferedImage,
) : BaseImageReader() {

    private val frame: ImageFrame = run {
        // For some reason some images have a greyscale type, even though they have color
        val converted = image.convertType(BufferedImage.TYPE_INT_ARGB)
        ImageFrame(converted, 1.milliseconds, Duration.ZERO)
    }
    private val frameDeferred: Deferred<ImageFrame> = CompletableDeferred(frame)

    override val size: Int = 1
    override val frameRate: Double = 1.0
    override val duration: Duration = 1.milliseconds
    override val frameDuration: Duration = 1.milliseconds
    override val width: Int = frame.content.width
    override val height: Int = frame.content.height
    override val loopCount: Int = 0

    override suspend fun readFrame(timestamp: Duration): ImageFrame = frame

    override fun createReversed(): ImageReader = this

    override fun iterator(): CloseableIterator<Deferred<ImageFrame>> = CloseableIterator.singleton(frameDeferred)

    override fun forEach(action: Consumer<in Deferred<ImageFrame>>) = action.accept(frameDeferred)

    override fun spliterator(): CloseableSpliterator<Deferred<ImageFrame>> =
        CloseableSpliterator.singleton(frameDeferred)

    override fun close() = Unit

    object Factory : ImageReaderFactory {
        override val supportedFormats: Set<String> = setOf(
            "bmp",
            "jpeg",
            "jpg",
            "wbmp",
            "png",
            "gif",
            "tif",
            "tiff",
        )

        override suspend fun create(input: DataSource): ImageReader = input.newStream().use {
            val image = withContext(Dispatchers.IO) {
                ImageIO.read(it) ?: throw IOException("Failed to read image")
            }
            JavaxImageReader(image)
        }
    }
}
